src/su/man/man1/sutwig.1
.TH SUTWIG 1 SU
.SH NAME
.B sutwig \-
full-featured wiggle trace program for data plotting
.SH SYNOPSIS
.nf
sutwig <stdin [optional parameters] | tube
.SH DESCRIPTION
.I Sutwig
produces wiggle trace byte codes.  The wiggle traces are put
on a graphics terminal screen by using a pen filter.
A full set of parameters is provided to allow publication
quality hard copy plots.
.P
The default plot size is designed
to fit in a 6"x8" hardcopy area, not to give a full-screen
plot.  A full-screen plot can be made by modifying the sizex
and sizet parameters.
.P
The shell program, tube(9SU), tries to deduce the proper pen from the
environment variable, TERM.
.P
The shell,
qtwig(9SU) evokes sutwig with no fill and no labels for the benefit of
slow graphics terminals.
.P
The shell,
qqtwig(9SU) evokes sutwig with no fill and no labels and windows
the data set (i.e selects a subset of traces) for the benefit of
slow graphics terminals.  This shell is of tutorial significance since
it illustrates how to allow shell files to handle arguments in getpar
format and thus how to write shells that act like su C programs (except
for execution speed).
.SH REQUIRED PARAMETERS
.TP 8
.I none
.SH OPTIONAL PARAMETERS
.TP 8
.B "LABELING PARAMETERS"
.TP
.I title
The string to be used as the plot title.  The default is no title.
.TP
.I label1
The string to be used for the vertical axis label.  The default is "Time."
.TP
.I label2
The string to be used for the horizontal axis label.
The default is "Trace."
.TP
.I tmin
The tic label for first vertical sample.  The default is the value of the
delrt (delay recording time) header word field on the first trace.
.TP
.I dt
The vertical tic labeling increment.  The default is the value of the
dt header word field (converted to seconds) on the first trace or
if this field is empty, .004 seconds.  Note that this parameter
affects only the labeling of the vertical scale and not the plot
itself.  The .004 default is dubious, but we find it convenient
for the common case when synthetic time traces are created and
the dt field is not filled.
.TP
.I xmin
The tic label for first horizontal sample.  The default is the value of the
tracl (trace number on line) header word field on the first trace.
.TP
.I dx
The horizontal tic labeling increment.  The default is the
difference of the values of the tracl field on the second and first
traces.
.TP
.I titlsz
The title print size; default 4.
.TP
.I lablsz
The label print size; default 4.
.TP
.I ticsz
The label print size; default 3.
.TP
.B MISCELLANEOUS PARAMETERS
.TP
.I ntict
The number of tics on the vertical axis; default 5.
.TP
.I nticx
The number of tics on the horizontal axis; default 4.
.TP
.I tlines
A flag for timing lines.  The default is 1 for having
timing lines, use 0 to suppress them.
.TP
.I fill
A flag to get positive fill.  The default value is 1 for having fill,
use 0 to suppress fill.
.TP
.I ltic
The length of a tic mark in inches; the default is 0.05.
.TP
.I plotfat
The line thickness of the traces.  For instance, plotfat=3 produces triply
thick lines.  Default is 1.
.TP
.I axisfat
The line thicknesses of the box and the tics.  Default is 1.
.TP
.B SIZE AND LOCATION PARAMETERS
.TP
.I sizet
The length of the vertical axis in inches.  The default is 6 inches.
.TP
.I sizex
The length of the horizontal axis in inches.  The default is 4.7 inches.
.TP
.I margint
The gap at the top and bottom between the box and the traces;
default is 0.1.
.TP
.I marginx
The side gap at the top and bottom between the box and the traces;
default is 0.1.
.TP
.I zerot
The space between the base of the plot and the bottom of screen in
inches.  The default is 1 inch.
.TP
.I zerox
The space between the left side of the plot and the left side of the
screen in inches.  The default is 0.6 inches.
.TP
.B DETAIL BOX PARAMETERS
.TP
.I db
A flag for including a detail box.  The default is 0 for no
detail box, use 1 to get a detail box.  (A "detail box" is a box drawn
on the plot, indicating the location of a subsequent blown up plot.
.TP
.I top
The vertical sample at the top of the detail box; default is nt/4, where
nt is the value of the ns (number of samples) field on the first trace.
.TP
.I bot
The vertical sample at the bottom of the detail box; default is nt/2.
.TP
.I left
The horizontal sample at the left of the detail box; default is ntr/2, where
ntr is the number of traces in the data set.
.TP
.I right
The horizontal sample at the right of the detail box; default is 3*ntr/4.
.TP
.I dbfat
The line thicknesses of the detail box and the tics; default is 4.
.TP
.B GAINING PARAMETERS
.TP
.I overlap
Controls the maximum deflection.  The default value is 2.
In terms of the gain parameter, "scale", the
maximum deflection in traces is overlap*scale.
If the gain parameters are defaulted (the usual case), then, scale=1,
and "overlap" is simply the maximum deflection measured in traces.
.TP
.I "gain parameters"
All gain parameters admitted by gainpkge(3SU) may be used.  The defaults
are: tpow=0.0, epow=0.0, gpow=1.0, agc=0, wagc=20, trap=0.0, clip=0.0,
qclip=1.0, qbal=1, pbal=0 and scale=1.0.
This implies that by default, the traces are balanced by the maximum
magnitude in the data set.
.SH DIAGNOSTICS
In addition to relaying messages from system call errors,
.I sutwig
complains and dies if the standard output is not redirected
(the output is byte code), with the message:
.TP 8
.I "must redirect or pipe byte code output"
.SH SEE ALSO
suwig(1) sutwig2(1) sutwig2v(1) supwig(1) gainpkge(3SU)
qtwig(9SU) qqtwig(9SU) tube(9SU)
.SH AUTHOR
Shuki, Chris, Jack.
.SH SOURCE
su/src/sutwig.c
.SH REVISION LEVEL
1.27


src/su/graphics/vplot/sutwig.c
/* SUTWIG: $Revision: 1.7 $ ; $Date: 90/11/10 13:21:38 $               */

#include "su.h"
#include "segy.h"
#include "vplot.h"

/*********************** self documentation ******************************/
string sdoc = "\
									\n\
SUTWIG - full-featured wiggle trace program for data plotting           \n\
									\n\
sutwig <stdin [optional parameters] | pen                               \n\
									\n\
Optional Parameters:                                                    \n\
....DATA...                                                             \n\
	tmin = tr.delrt         tic label for first time sample         \n\
	dt = tr.dt or .004      time sample increment                   \n\
	xmin = 1st tr.tracl     first trace location                    \n\
	dx = (1st-2nd)tr.tracl  trace sample increment                  \n\
....LABELING ...                                                        \n\
	title   = null          plot title                              \n\
	label1  = Time          time axis label                         \n\
	label2  = Trace         trace axis label                        \n\
	dtict = 0.0             time tic interval (0.0 for automatic)   \n\
	dticx = 0.0             trace tic interval (0.0 for automatic)  \n\
	ftict = tmin            1st numbered time tic (for dtict != 0)  \n\
	fticx = xmin            1st numbered trace tic (for dticx != 0) \n\
	titlsz = 9              title print size                        \n\
	lablsz = 9              label print size                        \n\
	ticsz  = 5              tic labeling print size                 \n\
....MISCELLANEOUS ...                                                   \n\
	tlines  = 1             flag for timing lines (=0 for none)     \n\
	fill    = 1             flag for positive fill (=0 for no fill) \n\
	ltic    = 0.05          length of tic mark (inches)             \n\
	plotfat = 0             line thickness of traces                \n\
	axisfat = 0             line thickness of box & tics            \n\
	hcopy   = 0             honors user/default parameters          \n\
				= 1   forces thesis format              \n\
....SIZE & LOCATION ...                                                 \n\
	sizet   = 6.0           length of t-axis (inches)               \n\
	sizex   = 4.7           length of x-axis (...)                  \n\
	margint = 0.1           top/bot gap between box and traces (...)\n\
	marginx = 0.1           side gap between box and traces (...)   \n\
	zerot   = 1.3           base of plot to bottom of screen (...)  \n\
	zerox   = 1.0           left of plot to left of screen (...)    \n\
....DETAIL BOX ...                                                      \n\
	db    = 0               flag for detail box (=1 for box)        \n\
	top   = nt/4            time sample at top of detail box        \n\
	bot   = nt/2            time sample at bottom of detail box     \n\
	left  = ntr/2           trace on left side of detail box        \n\
	right = 3*ntr/4         trace on right side of detail box       \n\
	dbfat = 4               line thickness for detail box           \n\
....GAINING ...                                                         \n\
	overlap = 2.0           max deflection (in traces) is           \n\
				overlap*scale                           \n\
									\n\
gain defaults (see sugain):                                             \n\
	tpow=0.0 epow=0.0 gpow=1.0 agc=0 wagc=0.2                       \n\
	trap=0.0 clip=0.0 qclip=1.0 qbal=1 pbal=0 scale=1.0             \n\
";
/**************** end self doc *******************************************/

/* Credits:
 *      SEP: Shuki
 *      CWP: Chris
 *
 */


/* Set plotting defaults */
#define TITLE	""
#define LABEL1  "Time"
#define LABEL2  "Trace"
#define FILL    1
#define OVERLAP 2.0
#define ZEROT   1.3
#define ZEROX   1.0
#define SIZET   6.0
#define SIZEX   4.5
#define MARGINT 0.1
#define MARGINX 0.1
#define DTICT   0.0;
#define DTICX   0.0
#define LTIC    0.05
#define PLOTFAT 0
#define AXISFAT 0
#define TITLSZ  9
#define LABLSZ  9
#define TICSZ   5
#define TLINES  1
#define DB      0
#define DBFAT   4

#define FONT	3


/* Set gain defaults (balance by maximum magnitude) */
#define TPOW    0.0
#define EPOW    0.0
#define GPOW    1.0
#define AGC     0
#define WAGC    0.2
#define TRAP    0.0
#define CLIP    0.0
#define QCLIP   1.0
#define QBAL    1       /* default is balance by maximum magnitude      */
#define PBAL    0
#define SCALE   1.0

segy tr;


main(int argc, char **argv)
{
	float *dataptr;         /* mega-vector of data from the segys   */
	float dt;               /* time sample rate (and tic increment) */
	float dx;               /* tic label increment for horiz axis   */
	float tmin;             /* minimum time (and tic label)         */
	float xmin;             /* minimum tic label for horiz axis     */
	int nt;                 /* time samples per trace (from tr.ns)  */
	int ntsize;             /* number of data bytes on a trace      */
	int ntr;                /* traces in input data (from gettr)    */
	int tracl2;             /* tracl of 2nd trace (for dx)          */
	int ndata;              /* bytes allocated for mega-vector      */
	string label1;          /* vertical axis label (default Time)   */
	string label2;          /* horiz axis label (default Trace)     */
	string title;           /* title on plot                        */
	void wigplot(           /* vertical wiggle plot                 */
		float *dataptr,
		int nt, int ntr,
		string title, string label1, string label2,
		float tmin, float dt, float xmin, float dx);


	/* Initialize */
	initargs(argc, argv);
	askdoc(1);


	/* Prevent bytes from spilling onto screen */
	if (isatty(STDOUT)) {
		err("must redirect or pipe byte code output");
	}


	/* Get info from first trace    */ 
	if (!gettr(&tr)) err("can't get first trace");
	nt = tr.ns;
	ntsize = nt * FSIZE;
	if (!fgetpar("xmin", &xmin))    xmin = tr.tracl;
	if (!fgetpar("tmin", &tmin))    tmin = tr.delrt/1000.0;

	if (!fgetpar("dt", &dt)) {
		if (tr.dt) {  /* is dt field set? */
			dt = tr.dt / 1000000.0;
		} else {                /* dt not set, assume 4 ms */
			dt = 0.004;
			warn("tr.dt not set, for labeling assume dt=%g", dt);
		}
	}
	

	/* Allocate block of memory for data float mega-vector */
	ndata = MAX(NFALLOC, nt); /* alloc at least one trace */
	dataptr = ealloc1float(ndata);


	/* Loop over input traces & put them into data mega-vector */
	ntr = 0;
	do {
		++ntr;
		if (ntr == 2) tracl2 = tr.tracl;/* needed for dx    */
		if (ntr*nt > ndata) {           /* need more memory */
			ndata += NFALLOC;
			dataptr = erealloc1float(dataptr, ndata);
		}
		bcopy(tr.data, dataptr + (ntr - 1)*nt, ntsize); 
	} while(gettr(&tr));


	if (!fgetpar("dx", &dx))         dx = tracl2 - xmin;


	/* Gain */
	gain(dataptr, TPOW, EPOW, GPOW, AGC, TRAP, CLIP, QCLIP,
			QBAL, PBAL, SCALE, tmin, dt, WAGC, nt, ntr);


	/* Plot getpars */
	if (!sgetpar("title",  &title))         title  = TITLE;
	if (!sgetpar("label1", &label1))        label1 = LABEL1;
	if (!sgetpar("label2", &label2))        label2 = LABEL2;


	/* Plot */
	warn("nt = %d  ntr = %d", nt, ntr);     
	wigplot(dataptr,nt, ntr, title, label1, label2, tmin, dt, xmin, dx);


	return EXIT_SUCCESS;
}



/* Wiggle plot subroutine for vertical plotting */
void wigplot(
	float *dataptr,
	int nt, int ntr,
	string title, string label1, string label2,
	float tmin, float dt, float xmin, float dx)
{
	char tval[8];           /* time value string                    */
	char xval[8];           /* x value string                       */
	string tfmt;            /* format string for time tic labels    */
	string xfmt;            /* format string for x tic labels       */
	float ltic;             /* length of tic marks (inches)         */
	float margint;          /* top/bot gap between box and traces   */
	float marginx;          /* side gap between box and traces      */
	float mt;               /* margint/scalet                       */
	float mx;               /* marginx/scalex                       */
	float overlap;          /* maximum trace overlap                */
	float scalet;           /* time axis scale                      */
	float scalex;           /* trace axis scale                     */
	float sizet;            /* length of t-axis (inches)            */
	float sizex;            /* length of x-axis (inches)            */
	float tpos;             /* temp for time position               */
	float xpos;             /* temp for trace position              */
	float zerot;            /* base of plot to bot. of screen       */
	float zerox;            /* left of plot to left of screen       */
	float tmax;             /* maximum t value on section           */
	float xmax;             /* maximum x value on section           */
	float tnum;             /* next t value to label                */
	float xnum;             /* next x value to label                */
	float tzero;            /* largest t value to round to zero     */
	float xzero;            /* largest x value to round to zero     */
	float dtict;            /* delta for time tics (0.0 automatic)  */
	float dticx;            /* delta for x tics (0.0 automatic)     */
	float ftict;            /* first time tic (dtict != 0.0)        */
	float fticx;            /* first x tic (dticx != 0.0)           */
	int axisfat;            /* line thickness of box & tics         */
	int db;                 /* 0=no detail box; 1=detail box        */
	int dbfat;              /* line thickness for detail box        */
	int fill;               /* fill flag                            */
	int hcopy;              /* hardcopy flag                        */
	int lablsz;             /* label print size                     */
	int left,top;           /* left side and top for detail box     */
	int ntict;              /* number of tics on time axis          */
	int nticx;              /* number of tics on trace axis         */
	int plotfat;            /* line thickness of traces             */
	int right, bot;         /* right side and bottom for detail box */
	int ticsz;              /* tic labeling print size              */
	int titlsz;             /* title print size                     */
	int tlines;             /* 1=timing lines (0=no timing lines)   */
	register int i;         /* counter                              */


	if (!igetpar("fill", &fill))            fill = FILL;

	if (!fgetpar("overlap", &overlap))      overlap = OVERLAP;

	if (!fgetpar("zerot", &zerot))          zerot = ZEROT;
	if (!fgetpar("zerox", &zerox))          zerox = ZEROX;

	if (!fgetpar("sizet", &sizet))          sizet = SIZET;
	if (!fgetpar("sizex", &sizex))          sizex = SIZEX;

	if (!fgetpar("margint", &margint))      margint = MARGINT;
	if (!fgetpar("marginx", &marginx))      marginx = MARGINX;

	if (!fgetpar("ltic", &ltic))            ltic = LTIC;

	if (!igetpar("plotfat", &plotfat))      plotfat = PLOTFAT;
	if (!igetpar("axisfat", &axisfat))      axisfat = AXISFAT;

	if (!igetpar("titlsz", &titlsz))        titlsz = TITLSZ;
	if (!igetpar("lablsz", &lablsz))        lablsz = LABLSZ;
	if (!igetpar("ticsz", &ticsz))          ticsz  = TICSZ;

	if (!igetpar("tlines", &tlines))        tlines = TLINES;

	/* Detail box parameters */
	if (!igetpar("db", &db))                db = DB;
	if (!igetpar("dbfat", &dbfat))          dbfat = DBFAT;
	if (!igetpar("left", &left))            left = ntr/2;
	if (!igetpar("top", &top))              top = nt/4;
	if (!igetpar("right", &right))          right = 3*ntr/4;
	if (!igetpar("bot", &bot))              bot = nt/2;

	if (!fgetpar("dtict", &dtict))          dtict = DTICT;
	if (!fgetpar("dticx", &dticx))          dticx = DTICX;
	if (dtict != 0.0)
		if (!fgetpar("ftict", &ftict))  ftict = tmin;
	if (dticx != 0.0)
		if (!fgetpar("fticx", &fticx))  fticx = xmin;

	if (!igetpar("hcopy", &hcopy))          hcopy = 0;
	if (hcopy) {
		titlsz = 3;
		lablsz = 3;
		ticsz = 2;
		sizet = 5.25;
		sizex = 4.4;
		zerot = 1.5;
		zerox = 1.;
	}

	vp_tfont(FONT, STRING, OVLY_NORMAL);

	scalet = -sizet/nt;
	scalex = sizex/MAX(ntr, 8);

	vp_scale (scalex, scalet);
	vp_orig (zerox, zerot + sizet);
	vp_fat (axisfat);

	mx = marginx/scalex;  mt = margint/scalet; 

	/* Title */
	vp_color (BLUE);  
	vp_tjust(TH_CENTER, TH_NORMAL);
	vp_utext (0.5*ntr, mt + 0.45/scalet, titlsz, 0, title);

	/* Labels */
	vp_color (BLUE);  
	vp_tjust(TH_CENTER, TH_NORMAL);
	vp_utext (-mx-0.3/scalex, 0.5*nt, lablsz, 90, label1); 
	vp_utext (0.5*ntr, nt-mt-0.6/scalet, lablsz, 0, label2);

	/* Axis box */
	vp_color (GREEN);  
	vp_umove ( -mx                  ,                  mt ); 
	vp_udraw ( -mx                  , (float) (nt-1) - mt );
	vp_udraw ( (float) (ntr-1) + mx , (float) (nt-1) - mt ); 
	vp_udraw ( (float) (ntr-1) + mx ,                  mt );
	vp_udraw ( -mx                  ,                  mt );

	/* Vertical axis tic marks  */
	tmax = tmin + (nt - 1)*dt;
	tzero = 0.0001 * (tmax - tmin);
	tfmt = "%1.4g";
  
	if (dtict == 0.0) {
		ntict = MAX(NINT(sizet), 5);
		scaxis(tmin, tmax, &ntict, &dtict, &ftict);
	}

	for (tnum = ftict; tnum <= tmax; tnum += dtict) {
		float t = (tnum - tmin)/dt;

		vp_umove ( -mx, (tnum-tmin)/dt );
		vp_where ( &xpos , &tpos );
		vp_draw ( xpos-ltic, tpos );
		if (tlines == 1) {
			vp_umove ( -mx , t );
			vp_udraw ( (float) (ntr-1) + mx , t ); 
		}       
		vp_umove ( mx + (float) (ntr-1) , t );
		vp_where ( &xpos , &tpos);
		vp_draw ( xpos + ltic , tpos );
		vp_where ( &xpos , &tpos );
		sprintf(tval, tfmt, tnum);
		if (ABS(tnum) < tzero) sprintf(tval, tfmt, 0.0);
		vp_tjust(TH_LEFT, TH_NORMAL);
		vp_text (xpos + 0.1, tpos, ticsz, 0, tval);
	}

	/* Horizontal axis tic marks  */
	xmax = xmin + (ntr - 1)*dx;
	xzero = 0.0001 * (xmax - xmin);
	xfmt = "%1.4g";
  
	if (dticx == 0.0) {
		nticx = MAX(NINT(sizex), 4);
		scaxis(xmin, xmax, &nticx, &dticx, &fticx);
	}
  
	for (xnum = fticx; xnum <= xmax; xnum += dticx) {
		float x = (xnum - xmin)/dx;

		vp_umove ( x, mt );
		vp_where ( &xpos , &tpos );
		vp_draw (xpos, tpos + 2*ltic);
		vp_umove ( x, -mt + (nt-1) );
		vp_where ( &xpos , &tpos );
		vp_draw (xpos, tpos - 2*ltic);
		vp_where ( &xpos , &tpos );
		sprintf(xval, xfmt, xnum);
		if (ABS(xnum) < xzero) sprintf(xval, xfmt, 0.0);
		vp_tjust(TH_CENTER, TH_NORMAL);
		vp_text (xpos, tpos - 0.22, ticsz, 0, xval);
	}

	/* Draw wiggle traces */
	vp_color (RED);  
	vp_fat (plotfat);
	vp_scale (scalex*overlap, scalet);
	for (i = 0; i < ntr; ++i) {
		vp_uorig (-(float) i / overlap, 0.0);
		vertvwig(dataptr + nt*i, nt, fill);
	}

	/* Detail box */
	vp_uorig (0,0);
	vp_scale (scalex, scalet);
	vp_color (GREEN);  
	if (db) { 
		vp_fat (dbfat); 
		vp_umove ( (float) (left -1) - .5 , (float) (top-1) );    
		vp_udraw ( (float) (right-1) + .5 , (float) (top-1) ); 
		vp_udraw ( (float) (right-1) + .5 , (float) (bot-1) ); 
		vp_udraw ( (float) (left -1) - .5 , (float) (bot-1) ); 
		vp_udraw ( (float) (left -1) - .5 , (float) (top-1) ); 
		vp_fat (axisfat); 
	} 

	vp_endplot ();
}
